geo_tag/geo_tag.py

```diff
@@ -0,0 +1,200 @@
+        if os.path.exists(filename):
+            cache_file = open(filename, 'rb')
+            self.cache = pickle.load(cache_file)
+            return
+        else:
+            feature = city_json.get('features')
+            for f in feature:
+                item = f.get('properties')
+                # store the city, state as key; (cityID, countyID, stateID, countyName) as value.
+                if item["name"] is not None and item["stateName"] is not None:
+                    _city_state = item["name"] + ", " + item["stateName"]
+                    # key, value
+                    self.cache[_city_state] = (item["cityID"], item["countyID"], item["stateID"], item["countyName"])
+
+                else:
+                    print("[ERROR] city.json data.")
+                    return
+
+            # store it using pickle
+            cache_file = open(filename, 'wb')
+            pickle.dump(self.cache, cache_file)
+
+    def get_coordinate(self):
+        coord = self.data.get('coordinates')
+        place = self.data.get('place')
+
+        if coord is not None:
+            self.coord = coord.get('coordinates')
+        else:
+            # 2. get the central point of the bounding_box
+            if place is not None:
+                bounding_box = place.get('bounding_box').get('coordinates')[0]
+                ''' Twitter has some wield format historically, though it still rectangle, 
+                but it is not always in (sw, se, ne, nw) order
+                '''
+                swLog = min(bounding_box[0][0], bounding_box[1][0], bounding_box[2][0], bounding_box[3][0])
+                swLat = min(bounding_box[0][1], bounding_box[1][1], bounding_box[2][1], bounding_box[3][1])
+                neLog = max(bounding_box[0][0], bounding_box[1][0], bounding_box[2][0], bounding_box[3][0])
+                neLat = max(bounding_box[0][1], bounding_box[1][1], bounding_box[2][1], bounding_box[3][1])
+
+                # AsterixDB is unhappy with this kind of point "rectangular"
+                if (swLog == neLog and swLat == neLat):
+                    swLog = neLog - 0.0000001
+                    swLat = neLat - 0.0000001
+
+                if (swLog > neLog or swLat > neLat):
+                    print("[ERROR] invalid coordinates of SW and NE.")
+                    return
+
+                # calculate the central point in the rectangle
+                self.coord = ((swLog+neLog) / 2, (swLat+neLat) / 2)
+
+    # help function
+    def extract_geo_tag_from_city_and_state(self, city_state_name, flag):
+        if city_state_name is None:
+            print("[ERROR] There is no city_name.")
+            return
+
+        if "," not in city_state_name:
+            print("[ERROR] parse the location name.")
+            return
+
+        state_of_city = city_state_name.split(",")[1]
+        target_state = str(state_of_city).strip()
+        full_state_name = self.abbrev_us_state[target_state]
```

Comment:
will this cause an IndexError? (not found in key)
```suggestion
        full_state_name = self.abbrev_us_state.get(target_state)
        if full_state_name is not None:
        
```